### PR TITLE
Provide guidance on ECHOConfig extension usage.

### PR DESCRIPTION
**Pull request by chris-wood**
This addresses #216.

cc @davidben, @bemasc, @mikebishop, @sftcd, @enygren

### CODE REVIEW

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -267,17 +267,20 @@ invalid if maximum_name_length is greater than 256.
...
-order, but there MUST NOT be more than one extension of the same type in the
-extensions block. An extension may be tagged as mandatory by using an extension
-type codepoint with the high order bit set to 1. A client which receives a
-mandatory extension they do not understand must reject the ECHOConfig content.
-
-Clients MUST parse the extension list and check for unsupported
-mandatory extensions. If an unsupported mandatory extension is
-present, clients MUST reject the ECHOConfig value.
+generating a ClientHello message. The purpose of the field is to provide room
+for additional features in the future. See {{config-extensions}} for guidance
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md:271]: features or protections

File: draft-ietf-tls-esni.md [hunk 2]
```diff
@@ -267,17 +267,20 @@ invalid if maximum_name_length is greater than 256.
...
+on what type of extensions are appropriate for this structure.
+
+The format is defined in {{RFC8446}}; Section 4.2. The same interpretation rules
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md:274]: Comma

File: draft-ietf-tls-esni.md [hunk 3]
```diff
@@ -934,6 +937,17 @@ IANA is requested to create an entry, echo_required(121) in the
 existing registry for Alerts (defined in {{!RFC8446}}), with the
 "DTLS-OK" column being set to "Y".
 
+# ECHOConfig Extension Guidance {#config-extensions}
+
+Any future information or hints that influence the outer ClientHello should be
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md:942]: SHOULD?

File: draft-ietf-tls-esni.md [hunk 4]
```diff
@@ -934,6 +937,17 @@ IANA is requested to create an entry, echo_required(121) in the
...
+specified as ECHOConfig extensions. This is primarily because the outer ClientHello
+exists only in support of ECHO. Namely, it is both an envelope for the encrypted
+inner ClientHello and enabler for authenticated key mismatch signals
+(see {{server-behavior}}). In contrast, the inner ClientHello is the true ClientHello
+used upon ECHO negotiation. Any information that influences the inner ClientHello,
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md:947]: Except, as we have discussed, the length, though I guess that is a part of the outer, too.

Comment (chris-wood) [reply draft-ietf-tls-esni.md:947]: Yep, I'm considering that part of the outer.

File: draft-ietf-tls-esni.md (hunk 1 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:271]: Isn't protection a feature?

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:274]: Will fix here and elsewhere!

File: draft-ietf-tls-esni.md (hunk 3 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:942]: I read them both the same, so that works for me.

**Comment by MikeBishop** [comment]
This looks fine, though I'll note that since an HTTPSSVC record is public and the outer ClientHello record is public, it's not unreasonable to pull values from there to put in the public fields since the observer already knows (or can get) them.  Restricting them to come only from the ECHOConfig seems unnecessary.

I think the key observation in this PR is that the outer ClientHello isn't _really_ offering to establish a TLS connection -- it's a not-crazy-looking wrapper for the actual ClientHello inside the ECHO extension, and the vehicle for the recovery flow.  The only things a server can legitimately do are respond to the inner ClientHello or do a recovery handshake.  Thus, most of the things you might set on the outer ClientHello are irrelevant; they can be chosen to look innocuous, chosen at random, or copied from the inner ClientHello as the implementation chooses.

**Comment by chris-wood** [comment]
@davidben, does this seem reasonable to you?

File: draft-ietf-tls-esni.md (hunk 1 above)
Comment (ekr) [reply draft-ietf-tls-esni.md:271]: functionality?

File: draft-ietf-tls-esni.md [hunk 5]
```diff
@@ -934,6 +937,17 @@ IANA is requested to create an entry, echo_required(121) in the
 existing registry for Alerts (defined in {{!RFC8446}}), with the
 "DTLS-OK" column being set to "Y".
 
+# ECHOConfig Extension Guidance {#config-extensions}
+
+Any future information or hints that influence the outer ClientHello SHOULD be
+specified as ECHOConfig extensions. This is primarily because the outer ClientHello
```
Comment (ekr) [inline draft-ietf-tls-esni.md:943]: Or, I guess, as ECHOConfigV2, right?

File: draft-ietf-tls-esni.md [hunk 6]
```diff
@@ -934,6 +937,17 @@ IANA is requested to create an entry, echo_required(121) in the
...
+exists only in support of ECHO. Namely, it is both an envelope for the encrypted
+inner ClientHello and enabler for authenticated key mismatch signals
+(see {{server-behavior}}). In contrast, the inner ClientHello is the true ClientHello
+used upon ECHO negotiation. Any information that influences the inner ClientHello,
+if needed, therefore belongs elsewhere, such as the corresponding HTTPSSVC
+record {{!HTTPSSVC}}.
+
```
Comment (ekr) [inline draft-ietf-tls-esni.md:950]: I'm not sure there is a need to be this directive here. Why not leave this for future work.

Also, hypothetically, suppose that the extension was like "make the nonce twice as long"

File: draft-ietf-tls-esni.md (hunk 1 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:271]: Functionality is a bit more general -- let's go with that!

File: draft-ietf-tls-esni.md (hunk 5 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:943]: Yeah, I'll add that: "as ECHOConfig extensions, or an entirely new version of ECHOConfig."

File: draft-ietf-tls-esni.md (hunk 6 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:948]: That's fair (and a good point). It is a bit... restrictive. I just removed the sentence.

**Pull request merged**